ADAL/src/request/ADAuthenticationRequest+Broker.m

```diff
@@ -189,65 +189,104 @@ + (ADAuthenticationResult *)processBrokerResponse:(NSURL *)response
     NSString *qp = [components percentEncodedQuery];
     //expect to either response or error and description, AND correlation_id AND hash.
     NSDictionary* queryParamsMap = [NSDictionary msidURLFormDecode:qp];
     
     if ([queryParamsMap valueForKey:MSID_OAUTH2_ERROR_DESCRIPTION])
     {
-        MSIDBrokerResponse *brokerResponse = [[MSIDBrokerResponse alloc] initWithDictionary:queryParamsMap error:nil];
-        return [ADAuthenticationResult resultFromBrokerResponse:brokerResponse];
+        // In the case where Intune App Protection Policies are required, the broker may send back the Intune MAM Resource token
+        NSMutableDictionary *brokerResponse = [[NSMutableDictionary alloc] initWithDictionary:queryParamsMap];
+        if (queryParamsMap[ADAL_BROKER_INTUNE_HASH_KEY] && queryParamsMap[ADAL_BROKER_INTUNE_RESPONSE_KEY])
+        {
+            ADAuthenticationError *intuneTokenError = nil;
+            NSDictionary *responseDictionary = @{ADAL_BROKER_RESPONSE_KEY:queryParamsMap[ADAL_BROKER_INTUNE_RESPONSE_KEY],
+                                                 ADAL_BROKER_HASH_KEY:queryParamsMap[ADAL_BROKER_INTUNE_HASH_KEY],
+                                                 ADAL_BROKER_MESSAGE_VERSION:queryParamsMap[ADAL_BROKER_MESSAGE_VERSION] ? queryParamsMap[ADAL_BROKER_MESSAGE_VERSION] : @1};
+
+            NSDictionary *decryptedIntuneTokenResponse = [ADBrokerKeyHelper decryptBrokerResponse:responseDictionary
+                                                                                    correlationId:correlationId
+                                                                                            error:&intuneTokenError];
+
+            NSError *tokenResponseError = nil;
+            MSIDBrokerResponse *intuneTokenResponse = [[MSIDBrokerResponse alloc] initWithDictionary:decryptedIntuneTokenResponse error:&tokenResponseError];
+
+            if (!keychainGroup)
+            {
+                MSID_LOG_WARN(nil, @"Failed to cache Intune token, unable to acquire keychain group.");
+            }
+            else if (tokenResponseError)
+            {
+                MSID_LOG_WARN(nil, @"Error parsing Intune token response");
+            }
+            else
+            {
+                
+                ADAuthenticationResult *intuneTokenResult = [ADAuthenticationResult resultFromBrokerResponse:intuneTokenResponse];
+                if (AD_SUCCEEDED != intuneTokenResult.status)
+                {
+                    MSID_LOG_WARN(nil, @"Failed to acquire Intune token.");
+                }
+                else
+                {
+                    if (intuneTokenResult.tokenCacheItem.userInformation.userId)
+                    {
+                        [brokerResponse setValue:intuneTokenResult.tokenCacheItem.userInformation.userId forKey:@"user_id"];
+                    }
+
+                    MSIDKeychainTokenCache *dataSource = [[MSIDKeychainTokenCache alloc] initWithGroup:keychainGroup];
```

Comment:
@jasoncoolmax, @unpluggedk FYI we should really split up this method as part of MSAL broker support as it has grown too big. I'm fine leaving it as is for this release.